cove_iati/lib/iati.py

```diff
@@ -447,3 +479,93 @@ def organisation_identifier_count(tree):
     unique_identifiers = list(set(identifiers))
 
     return len(unique_identifiers)
+
+
+ACTIVITY_PREFIX = '/iati-activities/iati-activity'
+
+
+def check_activity_org_refs(tree):
+    root = tree.getroot()
+    
+    publishers = requests.get("https://codelists.codeforiati.org/api/json/en/ReportingOrganisation.json").json()
+    registration_agency = requests.get("https://codelists.codeforiati.org/api/json/en/OrganisationRegistrationAgency.json").json()
+
+    publisher_codes = {publisher['code']: publisher for publisher in publishers['data']}
+
+    org_prefixes = {prefix['code']: prefix for prefix in registration_agency['data']}
+
+    found_publisher_orgs = {}
+    found_org_prefix = {}
+    not_found_orgs = {}
+
+    regex = re.compile('^[^\/\&\|\?]+$')
+
+    for activity in root.xpath(ACTIVITY_PREFIX):
+        reporting_org = activity.xpath('reporting-org/@ref')
+        iati_identifiers = activity.xpath('iati-identifier/text()')
+        iati_identifier = iati_identifiers[0] if len(iati_identifiers) else None
+
+        orgs_in_data = {
+            "Participating Org": activity.xpath('participating-org/@ref'),
+            "Transaction Provider": activity.xpath('transaction/provider-org/@ref'),
+            "Transaction Receiver": activity.xpath('transaction/receiver-org/@ref')
+        }
+        org_type_template = {key: 0 for key in orgs_in_data}
+
+        for org_type, orgs in orgs_in_data.items():
+            for org in orgs:
+                if org in reporting_org:
+                    continue
+
+                # get possible prefixes by initial substrings of ref. This is faster than checking ref
+                # against all prefixes in registry.
+                found_prefix = None
+                for possible_prefix in set(org[0:n] for n in range(4, len(org)+1)):  # no prefix shorther than 4
+                    if possible_prefix in org_prefixes:
+                        found_prefix = possible_prefix
+                        break
```

Comment:
Finding found_prefix happens even if it's not needed, because org is in publisher_codes

Reply:
But, the actual logic here looks good to me!